### PR TITLE
feat:tx support

### PR DESCRIPTION
**Pull request by constwz**
## Checklist
- [ ] Targeted PR against correct branch.
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Wrote unit tests.  
- [ ] Re-reviewed `Files changed` in the Github PR explorer.

### CODE REVIEW

File: database/database.go [hunk 1]
```diff
@@ -383,6 +379,14 @@ func (db *Impl) SaveCommitSignatures(ctx context.Context, signatures []*types.Co
 }
 
 func (db *Impl) SaveBucket(ctx context.Context, bucket *models.Bucket) error {
+	txAny := ctx.Value(common.UseTx(UseTransaction))
+	if tx, ok := txAny.(*Impl); ok && tx != nil {
```
Comment (irrun) [inline database/database.go:383]: a little weird, why not just set a bool value in context

Comment (constwz) [reply database/database.go:383]: If it is bool, how do I get the db instance that started the transaction

Comment (irrun) [reply database/database.go:383]: the func receiver?

Comment (constwz) [reply database/database.go:383]: If func receives, transactions are not required in all scenarios, so consider using the context tape

Comment (reneecok) [reply database/database.go:383]:
```
func (db *Impl) SaveBucket(ctx context.Context, bucket *models.Bucket) error {
	useTx := ctx.Value("use_transaction")
	tx := db
	if useTransaction, ok := useTx.(bool); ok && useTransaction {
		tx = tx.Begin(ctx)
	}

	err := tx.Db.Table((&models.Bucket{}).TableName()).Clauses(clause.OnConflict{
		Columns:   []clause.Column{{Name: "bucket_id"}},
		UpdateAll: true,
	}).Create(bucket).Error
	return err
}
```

**Pull request closed without merging**